### PR TITLE
Enable manual user activation

### PR DESCRIPTION
**Pull request by jdavcs**
Re #3208
Assumptions: confirmation not needed; single user activation only.
Refactoring: move user activation into user manager; update user controller.
TODO: selenium test (ideally, I think).

### CODE REVIEW

**Review by nsoranzo — APPROVED** [review]

**Comment by jdavcs** [comment]
Afterthoughts. 
1. The activation button is displayed regardless of the value of `app.config.user_activation_on`. Rationale: option might have been on, then turned off, with account(s) still needing activation.
2. I did not add a "deactivate user" option because I think it's not necessary. If, however, it could be helpful, pls let me know. Same for activating multiple users.

**Comment by martenson** [comment]
> Rationale: option might have been on, then turned off, with account(s) still needing activation.

If the activation is turned off, the flag is never checked -> I think the button should be hidden.


> I did not add a "deactivate user"

+1

**Comment by martenson** [comment]
I don't see the second commit in the list or in the files diff. 😕

**Comment by jdavcs** [comment]
> I don't see the second commit in the list or in the files diff. 😕

What second commit?

**Comment by martenson** [comment]
Oh that was just a reference, nevermind. 🤣

**Comment by jdavcs** [comment]
> If the activation is turned off, the flag is never checked -> I think the button should be hidden.

I was about to agree.. But then I thought about it; I don't think we should disable that button if the flag is not set. (1) If the feature is disabled, there will be no new accounts that are not activated - so the button will not be displayed for those. It will only be displayed for non-activate accounts, and non-active accounts can only be possible if the feature (a) is enabled, or (b) has been enabled previously. And in the latter, I think it's convenient to have a button to cleanup after unsetting the flag. And (2): the flag does not disable user *activation status* - it disables the *policy* that requires that users must confirm their email address, after which the system activates their account. User status exists regardless of that policy. The button is an admin override for that status. What do you think? (also, I'm not a galaxy admin, so I'm only guessing here)

**Comment by martenson** [comment]
I followed your rationale but suspect you might be overthinking it. I don't see a benefit in displaying control elements for features that are explicitly turned off.

**Comment by jdavcs** [comment]
> I followed your rationale but suspect you might be overthinking it. I don't see a benefit in displaying control elements for features that are explicitly turned off.

I might be. OK, I'll figure out how to make that change.

**Comment by jdavcs** [comment]
I can't see a better way to do this. In general, passing the app object may be useful in other scenarios down the road: the app holds the config, and grid elements may need to behave differently based on config settings. @martenson ? @nsoranzo : this has changed considerably since your approval.
EDIT: ping @dannon too - just noticed your reaction)

File: lib/galaxy/managers/users.py [hunk 1]
```diff
@@ -522,6 +522,11 @@ def send_subscription_email(self, email):
                 log.exception('Subscribing to the mailing list has failed.')
                 return "Subscribing to the mailing list has failed."
 
+    def activate(self, trans, user):
```
Comment (dannon) [inline lib/galaxy/managers/users.py]: We don't need trans in this function if you use `self.session()`

**Review by dannon — CHANGES_REQUESTED** [review]
I would not do this like this.

File: lib/galaxy/webapps/galaxy/controllers/admin.py [hunk 2]
```diff
@@ -136,22 +136,26 @@ def get_value(self, trans, grid, user):
...
                             target="top"),
         grids.GridOperation("Recalculate Disk Usage",
-                            condition=(lambda item: not item.deleted),
+                            condition=(lambda item, app: not item.deleted),
                             allow_multiple=False),
         grids.GridOperation("Generate New API Key",
+                            allow_multiple=False,
+                            async_compatible=True),
+        grids.GridOperation("Activate User",
+                            condition=(lambda item, app: app.config.user_activation_on and not item.active),
```
Comment (dannon) [inline lib/galaxy/webapps/galaxy/controllers/admin.py]: I would suggest not adding the option in this list, swapping the lambda back to a single argument without `app`, and then just conditionally add the GridOperation when the grid is initialized, based on the config at hand.

**Comment by dannon** [comment]
@ic4f Sorry, github wouldn't accept the review comment submission, and now it's attached all wonky.  Let me know if the suggestion isn't clear and I can get an example together

**Comment by jdavcs** [comment]
@dannon - no worries! I don't see the wonky :) thing, and your comments make sense. I wasn't at all sure adding the arg was the best/only possible way. Will change it. Thanks for reviewing!

**Comment by dannon** [comment]
@ic4f Ok, cool, just let me know if I can help :)  And, thanks for working on the suggestions!

**Comment by jdavcs** [comment]
@dannon Updated. Thanks - now I see it's done that way for other operations too.

**Comment by dannon** [comment]
@ic4f Much better, thank you.

**Comment by martenson** [comment]
Thanks @ic4f !

**Pull request merged**